src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,492 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
     }
 };
 
+class fetch_worker {
+public:
+    // Passed from the coordinator shard to fetch workers.
+    // Contains either references to objects local to the fetch worker shard or
+    // copies of data from the coordinator shard.
+    struct shard_local_fetch_context {
+        // True if the results of this fetch will be read on a foreign shard.
+        bool foreign_read;
+        size_t bytes_left;
+        // Specifies the minimum number of bytes this sub-fetch should read
+        // before returning.
+        size_t min_bytes;
+        // If set then the sub-fetch should return by the specified time_point .
+        std::optional<model::timeout_clock::time_point> deadline;
+        // The fetch sub-requests of partitions local to the shard this worker
+        // is running on.
+        std::vector<ntp_fetch_config> requests;
+
+        // References to services local to the shard this worker is running on.
+        // They are protected from deletion by the coordinator.
+        server& srv;
+        cluster::partition_manager& mgr;
+        ss::abort_source& as;
+    };
+
+    explicit fetch_worker(shard_local_fetch_context ctx)
+      : _ctx(std::move(ctx)) {}
+
+    struct worker_result {
+        std::vector<read_result> read_results;
+        // The total amount of bytes read across all results in `read_results`.
+        size_t total_size;
+    };
+
+    ss::future<worker_result> run() {
+        // Set the worker's abort source to be a child of the orchestrator's
+        // abort_source.
+        auto sub_opt = _ctx.as.subscribe([this]() noexcept {
+            _as.request_abort();
+            _completed_waiter_count.signal();
+        });
+
+        if (!sub_opt) {
+            // `_ctx.as` must've already been aborted. Query partitions once
+            // then return.
+            _as.request_abort();
+        }
+
+        auto result = co_await do_run();
```

Comment:
If do_run() throws we don't do the below, I guess the primary concern of which is that we don't do the abort/_waiter_gate.close(), so it seems the lifetime scoping can be violated?

Reply:
Maybe we should use ss::defer ot other RAII approach to guarantee it.

Reply:
Moving the abort and gate close to a `finally` block for `do_run` that should ensure that they run even if an exception is thrown.

Reply:
Sounds good (right RAII is not possible here as `close()` is async).